src/main/java/com/exasol/adapter/document/mapping/auto/SchemaInferencer.java

```diff
@@ -0,0 +1,79 @@
+package com.exasol.adapter.document.mapping.auto;
+
+import java.util.Optional;
+import java.util.logging.Logger;
+
+import com.exasol.adapter.document.edml.EdmlDefinition;
+import com.exasol.adapter.document.edml.MappingDefinition;
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This class automatically infers the schema for an {@link EdmlDefinition} in case the user did not specify a
+ * {@link EdmlDefinition#getMapping() mapping} when creating the virtual schema.
+ */
+public class SchemaInferencer {
+
+    private static final Logger LOG = Logger.getLogger(SchemaInferencer.class.getName());
+    private final SchemaFetcher schemaFetcher;
+
+    /**
+     * Create a new instance.
+     * 
+     * @param schemaFetcher the fetcher for retrieving the schema
+     */
+    public SchemaInferencer(final SchemaFetcher schemaFetcher) {
+        this.schemaFetcher = schemaFetcher;
+    }
+
+    /**
+     * If the given {@link EdmlDefinition} does not contain a {@link EdmlDefinition#getMapping() mapping}, this will
+     * return a new {@link MappingDefinition} with an automatically generated mapping based on the source. If a
+     * {@link EdmlDefinition#getMapping() mapping} is present, this will return the {@link EdmlDefinition} unmodified.
+     * 
+     * @param edmlDefinition the {@link EdmlDefinition} for which to infer the schema if it is missing.
+     * @return the updated {@link EdmlDefinition}.
+     */
```

Comment:
```suggestion
    /**
     * There are two ways to provide a mapping: 1) Either the {@link EdmlDefinition} contains an explicit mapping provided upfront and 2) infer the mapping from the {@link EdmlDefinition#getMapping() source data}.
     * If the given {@link EdmlDefinition} contains a {@link EdmlDefinition#getMapping() mapping}, the method will 
     * return this. Otherwise the method infers a mapping from the {@link EdmlDefinition#getMapping() source} and adds this to the {@link EdmlDefinition}. In both cases the method returns the {@link EdmlDefinition}.
     * 
     * @param edmlDefinition the {@link EdmlDefinition} from which to get or infer the mapping
     * @return {@link EdmlDefinition}, either unchanged or with added mapping
     * @throws IllegalStateExceptionin case mapping inference fails
     * @throws IllegalArgumentException in case the current VSD dialect does not support mapping inference
     */
```